homeassistant/components/http/forwarded.py

```diff
@@ -0,0 +1,165 @@
+      - If the number of elements in X-Forwarded-Proto does not equal 1 or
+        is equal to the number of elements in X-Forwarded-For, an HTTP 400
+        status code is thrown.
+      - If an empty X-Forwarded-Host is provided, a HTTP 400 status code is thrown.
+      - If an empty X-Forwarded-Proto is provided, or an empty element in the list,
+        a HTTP 400 status code is thrown.
+    """
+    #
+    @middleware
+    async def forwarded_middleware(request, handler):
+        """Process forwarded data by a reverse proxy."""
+        overrides = {}
+
+        # Handle X-Forwarded-For
+        forwarded_for = request.headers.getall(X_FORWARDED_FOR, [])
+        if not forwarded_for:
+            # No forwarding headers, continue as normal
+            return await handler(request)
+
+        # Ensure the IP of the connected peer is trusted
+        connected_ip = ip_address(request.transport.get_extra_info("peername")[0])
+        if not any(connected_ip in trusted_proxy for trusted_proxy in trusted_proxies):
+            _LOGGER.warning(
+                "Received X-Forwarded-For header from untrusted proxy %s, headers not processed",
+                str(connected_ip),
+            )
+            # Not trusted, continue as normal
+            return await handler(request)
+
+        # Multiple X-Forwarded-For headers
+        if len(forwarded_for) > 1:
+            _LOGGER.error("Too many headers for X-Forwarded-For: %s", forwarded_for)
+            raise HTTPBadRequest
+
+        # Process X-Forwarded-For from the right side (by reversing the list)
+        forwarded_for = list(reversed(forwarded_for[0].split(",")))
+        try:
+            forwarded_for = [ip_address(addr.strip()) for addr in forwarded_for]
+        except ValueError:
+            _LOGGER.error("Invalid IP address in X-Forwarded-For: %s", forwarded_for)
+            raise HTTPBadRequest
+
+        # Find the last trusted index in the X-Forwarded-For list
+        index = 0
+        for forwarded_ip in forwarded_for:
+            if any(forwarded_ip in trusted_proxy for trusted_proxy in trusted_proxies):
+                index += 1
+                continue
+            overrides["remote"] = str(forwarded_ip)
+            break
+
+        # If all the IP addresses are from trusted networks, take the left-most.
+        if "remote" not in overrides:
+            index = -1
+            overrides["remote"] = str(forwarded_for[-1])
+
+        # Handle X-Forwarded-Proto
+        forwarded_proto = list(reversed(request.headers.getall(X_FORWARDED_PROTO, [])))
+        if forwarded_proto:
+            if len(forwarded_proto) > 1:
+                _LOGGER.error(
+                    "Too many headers for X-Forward-Proto: %s", forwarded_proto
+                )
+                raise HTTPBadRequest
+
+            forwarded_proto = [
+                proto
+                for proto in (p.strip() for p in forwarded_proto[0].split(","))
+                if proto
```

Comment:
Should we drop this, just as we did with IPs ?

Reply:
I've explicitly added this case, because of your suggestion on the matching number of proto's. If a proto is empty, it is dropped, thus the count doesn't match.

Reply:
Maybe just bail out if find an occurrence with a separate error is a better idea.

Reply:
Yes, 👍 on being explicit.